cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -342,11 +357,12 @@ class DatasetWriterDirectoryQueue : public util::AsyncDestroyable {
   Make(util::AsyncTaskGroup* task_group,
        const FileSystemDatasetWriteOptions& write_options,
        DatasetWriterState* writer_state, std::shared_ptr<Schema> schema,
-       std::string dir) {
+       std::string directory, std::string prefix) {
     auto dir_queue = util::MakeUniqueAsync<DatasetWriterDirectoryQueue>(
-        std::move(dir), std::move(schema), write_options, writer_state);
+        std::move(directory), std::move(prefix), std::move(schema), write_options,
+        writer_state);
     RETURN_NOT_OK(task_group->AddTask(dir_queue->on_closed()));
-    dir_queue->PrepareDirectory();
+    dir_queue->PrepareDirectory(prefix);
```

Comment:
By this point you have already called `std::move` on `prefix` so that variable is no longer valid.  You get away with it when the prefix is small because C++'s small string optimization means that a move on a small string is a no-op.  However, with a large string, this could lead to errors.

Also, since we've already stored it in `prefix_` we can just use `prefix_` directly and `PrepareDirectory` doesn't need to take any arguments.

Reply:
Made the change.